File: elixir/lib/datagrout_conduit/prism.ex
defmodule DatagroutConduit.Prism do
  @moduledoc "Data transformation, charting, rendering, export, and type bridging."

  @doc """
  AI-driven data transformation (`data-grout/prism.refract`).

  ## Params

    * `"goal"` - Natural language transformation goal (required)
    * `"payload"` - Data to transform (required)
    * `"verbose"` - Include detailed trace
    * `"chart"` - Include a chart in output
  """
  @spec refract(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def refract(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "prism.refract", opts)
  end

  @doc """
  AI-driven charting (`data-grout/prism.chart`).

  ## Params

    * `"goal"` - Chart description (required)
    * `"payload"` - Data to chart (required)
    * `"format"` - Output format
    * `"chart_type"` - Chart type (bar, line, pie, etc.)
    * `"title"`, `"x_label"`, `"y_label"`, `"width"`, `"height"`
  """
  @spec chart(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def chart(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "prism.chart", opts)
  end

  @doc """
  Generate a document toward a natural-language goal (`data-grout/prism.render`).

  ## Params

    * `"goal"` - Natural language content description (required)
    * `"payload"` - Input data (optional)
    * `"format"` - Output format: markdown, html, pdf, json (default: markdown)
    * `"sections"` - Optional list of section specs
  """
  @spec render(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def render(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "prism.render", opts)
  end

  @doc """
  Convert content to another format without LLM (`data-grout/prism.export`).

  ## Params

    * `"content"` - Data or string to export (required)
    * `"format"` - Target format (required)
    * `"style"` - Optional styling options
    * `"metadata"` - Optional document metadata
  """
  @spec export(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def export(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "prism.export", opts)
  end

  @doc """
  Semantic type transformation (`data-grout/prism.focus`).

  ## Params

    * `"data"` - Data to transform (required)
    * `"source_type"` - Source type annotation (required)
    * `"target_type"` - Target type annotation (required)
    * `"source_annotations"` - Optional source schema hints
    * `"target_annotations"` - Optional target schema hints
    * `"context"` - Optional context string
  """
  @spec focus(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def focus(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "prism.focus", opts)
  end
end

File: python/src/datagrout/conduit/namespaces/prism.py
"""Prism namespace — data transformation, charting, rendering, and export."""

from typing import Any, Dict, List, Optional


class PrismNamespace:
    """Data transformation, charting, rendering, and type bridging."""

    def __init__(self, client: Any):
        self._client = client

    async def refract(
        self,
        goal: str,
        payload: Any,
        *,
        verbose: bool = False,
        chart: bool = False,
        **kwargs: Any,
    ) -> Any:
        """AI-driven data transformation (`data-grout/prism.refract`)."""
        self._client._warn_if_not_dg("prism.refract")
        self._client._ensure_initialized()
        params: Dict[str, Any] = {
            "goal": goal,
            "payload": payload,
            "verbose": verbose,
            "chart": chart,
            **kwargs,
        }
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/prism.refract", params)
        )

    async def chart(
        self,
        goal: str,
        payload: Any,
        **kwargs: Any,
    ) -> Any:
        """AI-driven charting (`data-grout/prism.chart`)."""
        self._client._warn_if_not_dg("prism.chart")
        self._client._ensure_initialized()
        params: Dict[str, Any] = {"goal": goal, "payload": payload, **kwargs}
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/prism.chart", params)
        )

    async def render(
        self,
        goal: str,
        payload: Optional[Any] = None,
        *,
        format: str = "markdown",
        sections: Optional[List[str]] = None,
        **kwargs: Any,
    ) -> Any:
        """Generate a document toward a natural-language goal (`data-grout/prism.render`)."""
        self._client._warn_if_not_dg("prism.render")
        self._client._ensure_initialized()
        params: Dict[str, Any] = {"goal": goal, "format": format, **kwargs}
        if payload is not None:
            params["payload"] = payload
        if sections is not None:
            params["sections"] = sections
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/prism.render", params)
        )

    async def export(
        self,
        content: Any,
        format: str,
        *,
        style: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        **kwargs: Any,
    ) -> Any:
        """Convert content to another format (`data-grout/prism.export`)."""
        self._client._warn_if_not_dg("prism.export")
        self._client._ensure_initialized()
        params: Dict[str, Any] = {"content": content, "format": format, **kwargs}
        if style is not None:
            params["style"] = style
        if metadata is not None:
            params["metadata"] = metadata
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/prism.export", params)
        )

    async def focus(
        self,
        data: Any,
        source_type: str,
        target_type: str,
        **kwargs: Any,
    ) -> Any:
        """Semantic type transformation (`data-grout/prism.focus`)."""
        self._client._warn_if_not_dg("prism.focus")
        self._client._ensure_initialized()
        params: Dict[str, Any] = {
            "data": data,
            "source_type": source_type,
            "target_type": target_type,
            **kwargs,
        }
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/prism.focus", params)
        )
